src/relayclient/RelayClient.ts

```diff
@@ -0,0 +1,264 @@
+import KnownRelaysManager, { EmptyFilter } from './KnownRelaysManager'
+import AccountManager from './AccountManager'
+import RelayedTransactionValidator from './RelayedTransactionValidator'
+import HttpWrapper from './HttpWrapper'
+import { GSNConfig, RelayClientConfig } from './GSNConfigurator'
+
+export const EmptyApprove: AsyncApprove = async (): Promise<string> => {
+  return Promise.resolve('0x')
+}
+
+export const GasPricePingFilter: PingFilter = (pingResponse, gsnTransactionDetails) => {
+  if (
+    pingResponse.MinGasPrice != null &&
+    gsnTransactionDetails.gasPrice != null &&
+    parseInt(pingResponse.MinGasPrice) > parseInt(gsnTransactionDetails.gasPrice)
+  ) {
+    throw new Error(`Proposed gas price: ${gsnTransactionDetails.gasPrice}; relay's MinGasPrice: ${pingResponse.MinGasPrice}`)
+  }
+}
+
+interface RelayingAttempt {
+  transaction?: Transaction
+  error?: Error
+}
+
+export interface RelayingResult {
+  transaction?: Transaction
+  pingErrors: Map<string, Error>
+  relayingErrors: Map<string, Error>
+}
+
+export default class RelayClient {
+  private readonly config: RelayClientConfig
+  private readonly web3: Web3
+  private readonly httpClient: HttpClient
+  private readonly contractInteractor: ContractInteractor
+  private readonly knownRelaysManager: KnownRelaysManager
+  private readonly accountManager: AccountManager
+  private readonly relayHub: Address
+  private readonly asyncApprove: AsyncApprove
+  private readonly relayedTransactionValidator: RelayedTransactionValidator
+  private readonly pingFilter: PingFilter
+
+  /**
+   * create a RelayClient library object, to force contracts to go through a relay.
+   * @param web3  - the web3 instance to use.
+   * @param httpClient
+   * @param contractInteractor
+   * @param knownRelaysManager
+   * @param accountManager
+   * @param relayedTransactionValidator
+   * @param pingFilter
+   * @param config options
+   * @param relayHub
+   * @param asyncApprove
+   */
+  constructor (
+    web3: Web3,
+    httpClient: HttpClient,
+    contractInteractor: ContractInteractor,
+    knownRelaysManager: KnownRelaysManager,
+    accountManager: AccountManager,
+    relayedTransactionValidator: RelayedTransactionValidator,
+    relayHub: Address,
+    pingFilter: PingFilter,
+    asyncApprove: AsyncApprove,
+    config: RelayClientConfig
+  ) {
+    this.config = config
+    this.web3 = web3
+    this.httpClient = httpClient
+    this.contractInteractor = contractInteractor
+    this.knownRelaysManager = knownRelaysManager
+    this.relayedTransactionValidator = relayedTransactionValidator
+    this.accountManager = accountManager
+    this.pingFilter = pingFilter
+    this.relayHub = relayHub
+    this.asyncApprove = asyncApprove
+  }
+
+  /**
+   * Create an instance of {@link RelayClient} with all default implementations of its dependencies.
+   * @param web3
+   * @param gsnConfig
+   */
+  static new (web3: Web3, gsnConfig: GSNConfig): RelayClient {
```

Comment:
Well, I don't have a good solution, but I do not expect anyone to use the `constructor` unless they really know what they are doing (consider it `private` undocumented API).
Cannot use default values to constructor because there is dependency between components. (`contractInteractor` being passed to both relay manager and transaction validator, etc.)
This can be improved a lot, but this was not my focus in this commit.